matsim/src/main/java/ch/sbb/matsim/routing/pt/raptor/SwissRailRaptorData.java

```diff
@@ -546,9 +546,17 @@ public CachingTransferProvider getTransferProvider(RTransfer transfer, CachingTr
      * synchronized in order to avoid that multiple quad trees for the very same stop filter attribute/value combination are prepared at the same time 
      */
 	public synchronized void prepareStopFilterQuadTreeIfNotExistent(String stopFilterAttribute, String stopFilterValue) {
+		// if stopFilterAttribute/stopFilterValue combination exists
+		// we do not have to do anything
+		if (stopFilterAttribute2Value2StopsQT.containsKey(stopFilterAttribute)
+			&& stopFilterAttribute2Value2StopsQT.get(stopFilterAttribute).containsKey(stopFilterValue))
+			return;
+		// otherwise we have to add this map
+		// only in case there is no attribute key
 		if (!stopFilterAttribute2Value2StopsQT.containsKey(stopFilterAttribute)) {
 			stopFilterAttribute2Value2StopsQT.put(stopFilterAttribute, new HashMap<>());
 		}
+		
```

Comment:
nice catch! This indeed looks like it was an oversight before, that the QuadTree was generated over and over.

The PR is okay for me, but as we're talking about performance, some additional suggestions to make it even faster:

At the moment, there are at least 2 accesses to the map, in some cases even 4: 1) `...containsKey(..)` on line 551, 2) `...get(..)` (line 552), 3) `!...containsKey(..)` (line 556) and 4) `...get(..)` at the bottom of the method. This could be replaced by a single `computeIfAbsent()` like this:

```java
// if stopFilterAttribute/stopFilterValue combination exists we do not have to do anything
Map<String, QuadTree<TransitStopFacility>> filteredQTs = 
        this.stopFilterAttribute2Value2StopsQT.computeIfAbsent(stopFilterAttribute, key -> new HashMap<>());
if (filteredQTs.containsKey(stopFilterValue))
    return;

....

filteredQTs.put(stopFilterValue, stopsQTFiltered);
```

This way, there are less map-access operations, and less if-statements which should lead to an additional (albeit small) speed up.

Reply:
Great, thanks, will include it.